external-crates/move/crates/move-regex-borrow-graph/src/regex.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt::{Debug, Display};
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct Regex<Lbl> {
```

Comment:
Working through this a bit, I think your life will be easier if you just support the full set of REs here, except only allowing `.*` instead of `r*`. It would save you from implementing all the variants you have (`Regex`, `Walk`, `Extension`, etc).

If you take that approach, you would have
```
enum Regex<Lbl> {
  Label(Lbl), EmptySet, Epsilon, DotStar
  Concat(r1: Regex<Lbl>, r2: Regex<Lbl>),
  Alternate(r1: Regex<Lbl>, r2: Regex<Lbl>)
}
```
This is the full RE language except that you have `DotStar` instead of `Star(r: Regex<Lbl>)`. With that change, the algorithm for `r.remove_prefix(s)` is a really straightforward match with like a dozen one-line cases.

Reply:
I am dumb and struggled to write that code. 

I think if you allow `Concat(.*, _)` we will fail to terminate in the abstract interpreter and will need some form of weakening at loop heads--which is not something we currently support. 
I don't think you gain anything today from this additional expressivity. 

To me the more interesting thing would be to rewrite `remove_prefix` to be `(Regex, Regex) -> [Regex]` instead of the two separate variants today `(Regex, Extension) -> [Regex]` and `(Extension, Regex) -> [Regex]`

Happy to take a PR after this lands if you want to poke around? :D